projects/fluid_simulation/sim.cpp

```diff
@@ -0,0 +1,435 @@
+    if (aspectRatio < 1.0) {
+        aspectRatio = 1.0 / aspectRatio;
+    }
+
+    int min = round(resolution);
+    int max = round(resolution * aspectRatio);
+
+    return (GetApp()->GetWindowWidth() > GetApp()->GetWindowHeight()) ? ppx::int2(max, min) : ppx::int2(min, max);
+}
+
+void FluidSimulation::InitTextures()
+{
+    ppx::int2 simRes = GetResolution(mConfig.simResolution);
+    ppx::int2 dyeRes = GetResolution(mConfig.dyeResolution);
+
+    // Generate all the textures.
+    mCheckerboardTexture = std::make_unique<Texture>(this, "checkerboard", GetApp()->GetWindowWidth(), GetApp()->GetWindowHeight(), GetApp()->GetSwapchain()->GetColorFormat());
+    mDisplayTexture      = std::make_unique<Texture>(this, "display", GetApp()->GetWindowWidth(), GetApp()->GetWindowHeight(), GetApp()->GetSwapchain()->GetColorFormat());
+    mDitheringTexture    = std::make_unique<Texture>(this, "fluid_simulation/textures/LDR_LLL1_0.png");
+    mDrawColorTexture    = std::make_unique<Texture>(this, "draw color", GetApp()->GetWindowWidth(), GetApp()->GetWindowHeight(), GetApp()->GetSwapchain()->GetColorFormat());
+    mDyeTexture[0]       = std::make_unique<Texture>(this, "dye[0]", dyeRes.x, dyeRes.y, kRGBA);
+    mDyeTexture[1]       = std::make_unique<Texture>(this, "dye[1]", dyeRes.x, dyeRes.y, kRGBA);
+    mVelocityTexture[0]  = std::make_unique<Texture>(this, "velocity[0]", simRes.x, simRes.y, kRG);
+    mVelocityTexture[1]  = std::make_unique<Texture>(this, "velocity[1]", simRes.x, simRes.y, kRG);
+
+    InitBloomTextures();
+    InitSunraysTextures();
+}
+
+void FluidSimulation::InitBloomTextures()
+{
+    ppx::int2 res = GetResolution(mConfig.bloomResolution);
+    mBloomTexture = std::make_unique<Texture>(this, "bloom", res.x, res.y, kRGBA);
+    PPX_ASSERT_MSG(mBloomTextures.empty(), "Bloom FBOs already initialized");
+    for (int i = 0; i < mConfig.bloomIterations; i++) {
+        uint32_t width  = res.x >> (i + 1);
+        uint32_t height = res.y >> (i + 1);
+        if (width < 2 || height < 2)
+            break;
+        mBloomTextures.emplace_back(std::make_unique<Texture>(this, std::string("bloom frame buffer[") + std::to_string(i) + "]", width, height, kRGBA));
+    }
+}
+
+void FluidSimulation::InitSunraysTextures()
+{
+    ppx::int2 res       = GetResolution(mConfig.sunraysResolution);
+    mSunraysTexture     = std::make_unique<Texture>(this, "sunrays", res.x, res.y, kR);
+    mSunraysTempTexture = std::make_unique<Texture>(this, "sunrays temp", res.x, res.y, kR);
+}
+
+void FluidSimulation::AddTextureToInitialize(Texture* texture)
+{
+    mTexturesToInitialize.push_back(texture);
+}
+
+void FluidSimulation::GenerateInitialSplat()
+{
+    for (const auto& texture : mTexturesToInitialize) {
+        ScheduleDR(mColor->GetDR(texture, ppx::float4(0.0, 0.0f, 0.0f, 1.0f)));
+    }
+    mTexturesToInitialize.clear();
+
+    MultipleSplats(Random().UInt32() % 20 + 5);
```

Comment:
Probably a good idea to improve the `Random` class to add methods to return integers in ranges

Reply:
Not now.  File an enhancement request please?  Thanks.